### PR TITLE
perf(web3-indexer): Store block faster 

### PR DESCRIPTION
**Pull request by classicalliu**
Fetch and store transactions & logs in batch , be more faster if block has many transactions and logs.

> **Note**: This PR requires [web3-indexer-db migration](https://github.com/nervosnetwork/godwoken-web3/blob/7ab6a6ab4ccde7daa4e203fc06684ffdb0646001/packages/api-server/migrations/20220701114448_remove_foreign_keys.ts).

### CODE REVIEW

**Comment by Flouse** [comment]
> **Note**: This PR requires [web3-indexer-db migration](https://github.com/nervosnetwork/godwoken-web3/blob/7ab6a6ab4ccde7daa4e203fc06684ffdb0646001/packages/api-server/migrations/20220701114448_remove_foreign_keys.ts).

cc @RetricSu

File: crates/indexer/src/indexer.rs [hunk 1]
```diff
@@ -87,364 +104,415 @@ impl Web3Indexer {
...
+            ids_len
+        );
+
+        let scripts = batch_account_id_to_script(&self.godwoken_async_client, ids.clone()).await?;
+
+        let mut hashmap = HashMap::<u32, Option<Script>>::new();
+        scripts.into_iter().enumerate().for_each(|(i, value)| {
+            let id = ids[i];
+            hashmap.insert(id, value);
+        });
```
Comment (keroro520) [inline crates/indexer/src/indexer.rs:433]: [`zip`](https://doc.rust-lang.org/1.61.0/core/iter/fn.zip.html) may be more fitable here.

File: crates/indexer/src/indexer.rs [hunk 2]
```diff
@@ -87,364 +104,415 @@ impl Web3Indexer {
...
+        let mut total_gas_used: u128 = 0;
+        for txs in txs_slice {
+            let l2_transaction_with_logs_vec = txs
+                .into_par_iter()
+                .map(|tx| {
+                    self.filter_single_transaction(
+                        tx,
+                        block_number,
+                        block_hash,
+                        0,
```
Comment (keroro520) [inline crates/indexer/src/indexer.rs:474]: `tx_index` always be `0`?

File: crates/indexer/src/insert_l2_block.rs [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                let mut logs_query_builder: QueryBuilder<Postgres> = QueryBuilder::new(
+                    "INSERT INTO logs
+                    (transaction_id, transaction_hash, transaction_index, block_number, block_hash, address, data, log_index, topics)"
+                );
+
+                // Get transaction id from preview insert returning
+                logs_query_builder.push_values(db_logs, |mut b, log| {
+                    // transaction_id in log is transaction_id_index now
+                    let transaction_id_index = log.transaction_id as usize;
+                    let transaction_id = tx_ids[transaction_id_index];
```
Comment (keroro520) [inline crates/indexer/src/insert_l2_block.rs:251]: Can we use `let transaction_id = tx_ids[log.transaction_index];` here?

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: crates/indexer/src/indexer.rs (hunk 2 above)
Comment (classicalliu) [reply crates/indexer/src/indexer.rs:474]: Tx index unknown here, should update later: https://github.com/nervosnetwork/godwoken-web3/blob/refactor-sync/crates/indexer/src/indexer.rs#L485

File: crates/indexer/src/indexer.rs [hunk 4]
```diff
@@ -48,6 +56,7 @@ impl Web3Indexer {
         let mut allowed_eoa_hashes = HashSet::default();
         allowed_eoa_hashes.insert(eth_account_lock_hash);
         let godwoken_rpc_client = GodwokenRpcClient::new(gw_rpc_url);
+        let godwoken_async_client = GodwokenAsyncClient::with_url(gw_rpc_url).unwrap(); // TODO:
```
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/src/indexer.rs:59]: Deal error with `.unwrap()` here

File: crates/indexer/src/insert_l2_block.rs (hunk 3 above)
Comment (classicalliu) [reply crates/indexer/src/insert_l2_block.rs]: ok

File: crates/indexer/src/indexer.rs (hunk 1 above)
Comment (classicalliu) [reply crates/indexer/src/indexer.rs]: update

File: crates/indexer/src/indexer.rs (hunk 2 above)
Comment (keroro520) [reply crates/indexer/src/indexer.rs:476]: So this parameter 'tx_index' seems useless, is it? Let's remove this parameter from the function declaration?

**Review by keroro520 — APPROVED** [review]

File: crates/indexer/src/indexer.rs (hunk 2 above)
Comment (classicalliu) [reply crates/indexer/src/indexer.rs]: Updated

File: crates/indexer/src/indexer.rs [hunk 5]
```diff
@@ -87,364 +104,417 @@ impl Web3Indexer {
...
+
+        // begin db transaction
+        let pool = &*POOL;
+        let mut pg_tx = pool.begin().await?;
+
+        let mut tx_index_cursor: u32 = 0;
+
+        let mut cumulative_gas_used: u128 = 0;
+        let mut total_gas_limit: u128 = 0;
+        let mut total_gas_used: u128 = 0;
```
[the remark this answers is not in the document]
Comment (classicalliu) [reply crates/indexer/src/indexer.rs]: Seems has the same value, now updated to use only `cumulative_gas_used`

**Review by keroro520 — APPROVED** [review]

**Review by Flouse — APPROVED** [review]

**Review by RetricSu — APPROVED** [review]

**Comment by RetricSu** [comment]
please rebase and fix the conflicts

**Comment by classicalliu** [comment]
> please rebase and fix the conflicts @classicalliu

Done

**Pull request merged**